### PR TITLE
Lower bound priority

### PR DESCRIPTION
**Pull request by morcos**
I think this is a good compromise with regards to priority for 0.12 release. It is a calculation of priority in which only inputs that are in the chain at the time the transaction enters the mempool are aged.

It is a minimal amount of code and changes the `CTxMemPoolEntry::GetPriority` function to return a lower bound of the correctly computed priority as opposed to an upper bound.  This would allow the fast mining code in #6898 to use a priority calculation that at least ages the originally in chain inputs as opposed to using only starting priority.  
I believe this better strikes the right balance between preserving the ability to meet previous policy objectives and adding too much complication to the code for a concept that is deprecated.  I plan to close #6357 in favor of this.

One caveat: In the event of a reorg, if one of the originally in-chain inputs is now no longer in chain (or has been moved to a different height) the calculation will no longer be a strict lower bound on priority because it will not take this into account.  

It should be kept in mind though, that this is a clear improvement on the current calculation.

(Built on top of #7007 so the tests pass)

(EDIT: For anyone that has already reviewed #6357, this is almost a strict subset)

### CODE REVIEW

**Comment by sdaftuar** [comment]
utACK

**Comment by luke-jr** [comment]
What are the benefits of this over #6357 ?

**Comment by sipa** [comment]
It's simpler.

File: src/coins.cpp [hunk 1]
```diff
@@ -239,8 +240,9 @@ double CCoinsViewCache::GetPriority(const CTransaction &tx, int nHeight) const
         const CCoins* coins = AccessCoins(txin.prevout.hash);
         assert(coins);
         if (!coins->IsAvailable(txin.prevout.n)) continue;
-        if (coins->nHeight < nHeight) {
+        if (coins->nHeight <= nHeight) {
```
Comment (petertodd) [inline src/coins.cpp]: Why was this line changed?

Comment (petertodd) [inline src/coins.cpp]: Oh wait,  I get it, never mind.

File: src/txmempool.h [hunk 2]
```diff
@@ -78,15 +79,20 @@ class CTxMemPoolEntry
...
-                    int64_t _nTime, double _dPriority, unsigned int _nHeight, bool poolHasNoInputsOf = false);
+                    int64_t _nTime, double _entryPriority, unsigned int _entryHeight,
+                    bool poolHasNoInputsOf, CAmount _inChainInputValue);
     CTxMemPoolEntry(const CTxMemPoolEntry& other);
 
     const CTransaction& GetTx() const { return this->tx; }
+    /**
+     * Fast calculation of lower bound of current priority as update
+     * from entry priority.  Only originally in-chain inputs will age
+     */
```
Comment (petertodd) [inline src/txmempool.h]: "Only originally in-chain inputs will age" <- looks like you're missing something here.

**Comment by petertodd** [comment]
utACK, modulo comment nit.

File: src/txmempool.h (hunk 2 above)
Comment (morcos) [inline src/txmempool.h]: I'm happy to make it clearer, but that was the intended wording except for perhaps a missing ".".

Comment (petertodd) [inline src/txmempool.h]: How about you change it to "Only inputs that were originally in-chain will age."

**Comment by jtimon** [comment]
utACK

**Comment by dcousens** [comment]
Is it worth changing this in light of the eventual removal of priority?
If it is, ACK.

**Comment by sdaftuar** [comment]
Added a commit to address @petertodd's comment nit.

@dcousens Yes I think this change is needed for 0.12, since all the priority code won't be eliminated yet.  This seems like the best compromise for 0.12.

**Comment by jtimon** [comment]
In any case, merging this makes other mempool related PRs more readable, it's small and it doesn't preclude the future elimination of the priority code (and there's currently no rebased version of a priority-removing PR that I know of). So even if I'm very much in favor of directly removing the priority, I would be happy to see this merged soon.

I think some squashing could be done here though, at the very least for the comment fixup commit.

File: src/coins.cpp (hunk 1 above)
Comment (sipa) [inline src/coins.cpp]: Good catch, but indeed :)

**Comment by sipa** [comment]
utACK

**Comment by morcos** [comment]
rebased

File: src/txmempool.h [hunk 3]
```diff
@@ -63,9 +63,10 @@ class CTxMemPoolEntry
     size_t nModSize; //! ... and modified size for priority
     size_t nUsageSize; //! ... and total memory usage
     int64_t nTime; //! Local time when entering the mempool
-    double dPriority; //! Priority when entering the mempool
```
Comment (jtimon) [inline src/txmempool.h]: What was the motivation to bike-sheed dPriority and nHeight?

Comment (morcos) [inline src/txmempool.h]: This was originally the first part of the changes necessary for #6357 (now #7149) 
It was necessary to distinguish between the two stored priorities and heights.  The original "entry" priority and height and the latest calculated "cached" priority and the height that was calculated at.

It seemed logical to me to leave those changes in given that we didn't know whether the second half would be merged or not and given that they are more accurate names anyway.

Comment (jtimon) [inline src/txmempool.h]: > It was necessary to distinguish between the two stored priorities and heights. The original "entry" priority and height and the latest calculated "cached" priority and the height that was calculated at.

I guess if the rename had been introduced when it was necessary instead of here I wouldn't have complained.

> It seemed logical to me to leave those changes in given that we didn't know whether the second half would be merged or not and given that they are more accurate names anyway.

I'm confused, how are CTxMemPoolEntry::nHeight and CTxMemPoolEntry::dPriority "less accurate" than CTxMemPoolEntry::entryHeight and CTxMemPoolEntry::entryPriority ?
